tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,499 @@
+            .finish()
+    }
+}
+
+impl<B, ResBody> ValidateRequest<B> for AcceptHeader<ResBody>
+where
+    ResBody: Body + Default,
+{
+    type ResponseBody = ResBody;
+
+    fn validate(&mut self, req: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>> {
+        if !req.headers().contains_key(header::ACCEPT) {
+            return Ok(());
+        }
+        if req
+            .headers()
+            .get_all(header::ACCEPT)
+            .into_iter()
+            .flat_map(|header| {
+                header
+                    .to_str()
+                    .ok()
+                    .into_iter()
+                    .flat_map(|s| s.split(",").map(|typ| typ.trim()))
+            })
+            .any(|h| {
+                let value = self.header_value.to_str().unwrap();
+                let primary = format!("{}/*", value.split("/").nth(0).unwrap());
+                h == "*/*" || h == primary || h == value
+            })
+        {
+            return Ok(());
+        }
+        let mut res = Response::new(ResBody::default());
+        *res.status_mut() = StatusCode::NOT_ACCEPTABLE;
+        Err(res)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[allow(unused_imports)]
+    use super::*;
+    use http::header;
+    use hyper::Body;
+    use tower::{BoxError, ServiceBuilder, ServiceExt};
+
+    #[tokio::test]
+    async fn valid_accept_header() {
+        let mut service = ServiceBuilder::new()
+            .layer(ValidateRequestHeaderLayer::accept("application/json"))
+            .service_fn(echo);
+
+        let request = Request::get("/")
+            .header(header::ACCEPT, "application/json")
+            .body(Body::empty())
+            .unwrap();
+
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+
+        assert_eq!(res.status(), StatusCode::OK);
+    }
+
+    #[tokio::test]
+    async fn valid_accept_header_accept_all_json() {
+        let mut service = ServiceBuilder::new()
+            .layer(ValidateRequestHeaderLayer::accept("application/json"))
+            .service_fn(echo);
+
+        let request = Request::get("/")
+            .header(header::ACCEPT, "application/*")
+            .body(Body::empty())
+            .unwrap();
+
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+
+        assert_eq!(res.status(), StatusCode::OK);
+    }
+
+    #[tokio::test]
+    async fn valid_accept_header_accept_all() {
+        let mut service = ServiceBuilder::new()
+            .layer(ValidateRequestHeaderLayer::accept("application/json"))
+            .service_fn(echo);
+
+        let request = Request::get("/")
+            .header(header::ACCEPT, "*/*")
+            .body(Body::empty())
+            .unwrap();
+
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+
+        assert_eq!(res.status(), StatusCode::OK);
+    }
+
+    #[tokio::test]
+    async fn invalid_accept_header() {
```

Comment:
Perhaps a test checking that `application/dummy` is not accepted when `application/json` is used?